libtiledbvcf/src/vcf/vcf_v4.h

```diff
@@ -0,0 +1,200 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#ifndef TILEDB_VCF_VCF_V4_H
+#define TILEDB_VCF_VCF_V4_H
+
+#include <htslib/hts.h>
+#include <htslib/synced_bcf_reader.h>
+#include <htslib/vcf.h>
+#include <htslib/vcfutils.h>
+#include <algorithm>
+#include <cstdio>
+#include <iostream>
+#include <map>
+#include <queue>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "utils/utils.h"
+#include "vcf/htslib_value.h"
+#include "vcf/region.h"
+#include "vcf/vcf_utils.h"
+
+namespace tiledb {
+namespace vcf {
+
+/**
+ * Class wrapping a BCF/VCF file to allow iteration over records.
+ */
+class VCFV4 {
+ public:
+  VCFV4();
+  ~VCFV4();
+
+  VCFV4(VCFV4&& other) = delete;
+  VCFV4(const VCFV4&) = delete;
+  VCFV4& operator=(const VCFV4&) = delete;
+  VCFV4& operator=(VCFV4&&) = delete;
+
+  /**
+   * Opens the specified VCF or BCF file and load the header.
+   *
+   * @param file Path to VCF file.
+   * @param index_file Path to VCF index file. If empty, HTSlib will attempt
+   *   to locate the index file automatically.
+   */
+  void open(const std::string& file, const std::string& index_file = "");
+
+  /** Closes the VCF file and invalidates the iterator. */
+  void close();
+
+  /** Returns true if the file is open. */
+  bool is_open() const;
+
+  /**
+   * Sets the current iterator position to the first record at or after the
+   * given region. Getting `curr_rec` after seek will return the first record in
+   * the region.
+   *
+   * @param contig_name Name of contig to seek to
+   * @param pos 0-based starting position within contig to seek to
+   * @return True if there is at least one record in the given region.
+   */
+  bool seek(const std::string& contig_name, uint32_t pos);
+
+  /**
+   * Returns true if the given contig has any records in the VCF file.
+   */
+  bool contig_has_records(const std::string& contig_name) const;
+
+  /** Returns the header instance of the currently open file. */
+  bcf_hdr_t* hdr() const;
+
+  /**
+   * Returns the next buffered record but does not remove it from the queue.
+   * Note you must first call seek() to initialize the iterator position.
+   *
+   * @return null if all records have been read.
+   */
+  SafeSharedBCFRec front_record();
+
+  /**
+   * Removes the next buffered record from the internal queue. This does
+   * not invalidate records returned from `front_record`.
+   */
+  void pop_record();
+
+  /**
+   * As an optimization, records returned from `next()` may be passed
+   * to this routine for re-use to reduce memory allocation overhead.
+   *
+   * @param record The processed record to return into the allocation pool.
+   */
+  void return_record(SafeSharedBCFRec record);
```

Comment:
I don't see this used anywhere, but it could be `SafeSharedBCFRec&&` to avoid a copy if the caller is done with the `record`.

Reply:
It is used here: https://github.com/TileDB-Inc/TileDB-VCF/blob/dd3b459f4f682f6c02c9ed8e0ffab87c84d2f822/libtiledbvcf/src/write/writer_worker_v4.cc#L155